game-core/src/main/java/games/strategy/triplea/attachments/TriggerAttachment.java

```diff
@@ -1342,6 +1356,53 @@ private static void placeUnits(final TriggerAttachment t, final Territory terr,
     bridge.addChange(change);
   }
 
+  /**
+   * Test if we are resetting/clearing the variable first, and if so, remove the leading "-reset-" or "-clear-"
+   * from the new value.
+   */
+  public static Tuple<Boolean, String> getClearFirstNewValue(final String preNewValue) {
+    final boolean clearFirst = preNewValue.length() > 0
+        && (preNewValue.startsWith(PREFIX_CLEAR) || preNewValue.startsWith(PREFIX_RESET));
+    final String newValue = clearFirst
+        ? preNewValue.replaceFirst(PREFIX_CLEAR, "").replaceFirst(PREFIX_RESET, "")
+        : preNewValue;
+    return Tuple.of(clearFirst, newValue);
+  }
+
+  /**
+   * Returns the property change as well as history start event message.
+   *
+   * <p>
+   * No side-effects.
+   * </p>
+   */
+  static Optional<Tuple<Change, String>> getPropertyChangeHistoryStartEvent(
+      final TriggerAttachment triggerAttachment,
+      final DefaultAttachment propertyAttachment, final String propertyName,
+      final Tuple<Boolean, String> clearFirstNewValue,
+      final String propertyAttachmentName, final Named attachedTo) {
+    final boolean clearFirst = clearFirstNewValue.getFirst();
+    final String newValue = clearFirstNewValue.getSecond();
+
+    final boolean isValueTheSame = newValue.equals(propertyAttachment.getRawPropertyString(propertyName));
+
+    if (!isValueTheSame) {
```

Comment:
When you look at my proposal for `getClearFirstValue` this variable seems oddly familiar.
I haven't checked, but I'd assume that `isValueTheSame != clearFirst` will always be true.
If this is actually the case there would be no need to return `clearFirst `in the first place, thus simplifying the code.

Reply:
It is actually AFAICT different, when `isValueTheSame` is false and `newValue` is empty, `clearFirst` being true or false will cause different behaviour.

Reply:
I'm not sure if I understand you correctly.
I'll try to tell you about my thought process, so you can hopefully tell me where my assumption is wrong.

I believe `propertyAttachment.getRawPropertyString(propertyName)` returns the original value (with -clear- or -reset- prefixes, before the value gets passed to `getClearFirstNewValue`).
Therefore `isValueTheSame` will only be true if no such prefix that will be trimmed in `getClearFirstNewValue` exists for this property, but clearFirst is true only when such a prefix exists.

So the only reason why `isValueTheSame != clearFirst` wouldn't be true all of the time is that `propertyAttachment.getRawPropertyString(propertyName)` doesn't actually return what I'm expecting it to return. In this case my question would be: Was is this call actually returning?

I hope that makes sense somehow.

Reply:
I am uncertain, but after looking into it I believe that the design is as follows:

- The `clearFirstNewValue` is derived from a property, and that property is a part of the `TriggerAttachment` instance. This would in the XML for instance be: `<option name="playerProperty" value="productionPerXTerritories" count="-reset-4:conscript_s"/>` (taken from Greyhawk Wars). In this example, it contains the `-reset-` instruction part. You can see it gets this from the `TriggerAttachment` instance with `t.getPlayerProperty()` in the middle loop in `triggerPlayerPropertyChange`.
- The "property" attachment (or the attachment whose property we want to change once the trigger attachment triggers) with parameter name `propertyAttachment` (which I am not certain I named well - a better name might be "target" attachment/`targetAttachment` instead of `propertyAttachment`? Might it make sense to rename in the changes?), on the other hand, is not part of the `TriggerAttachment` instance, but is instead gotten from the `PlayerId` with variable name `player`: `PlayerAttachment.get(player, t.getPlayerAttachmentName().getSecond())` (do note that it first gets the `PlayerAttachment::get` from the `Map` in the refactored code).
- The "property"/"target" attachment(s) holds the actual value, and the trigger attachment contains the instructions for how to modify that value once the trigger attachment is triggered. The "property"/"target" attachment(s) will not hold the "-reset-" instruction part, for that instruction is only contained in the trigger attachment and only used when the trigger attachment triggers.

Reply:
Thanks for the clarification, I think I get it now.